### DISCUSSION TITLE
[Pester 5] Call "base" mock from another mock - a good idea?

### DISCUSSION THREAD

**Discussion by bozho** [pester/Pester#2179] [Q&A] [answered]
I'm doing a bunch of unit tests for module functions that are essentially 1-to-1 wrappers for REST method calls. One requirement is that we don't want request URLs to show up on VERBOSE stream, so the unit tests check that `Invoke-RestMethod` is always called with an explicit `-Verbose:$false` parameter.

For that purpose, I've written a small reusable mock (gets dot-sourced in each unit test file):
```powershell
$MockForceNonVerboseInvokeRestMethod = {
    if(!$PesterBoundParameters.ContainsKey("Verbose") -or ($PesterBoundParameters.Verbose -ne $false)) { throw [System.ArgumentException] "Invoke-RestMethod must be called with explicit -Verbose:`$false parameter." }
    return @(
        @{
            data = @{}
        }
    )
}
```

This works fine.

Now, we have a few module functions that wrap several `Invoke-RestMethod` calls (e.g. get some data, do something with it, `PUT` it back), which means I need custom mocks for those calls (need to return some mock data).

While it's simple to copy that one line from our "default" `Invoke-RestMethod` mock, I've also tried this:
```powershell
Mock -ModuleName $moduleName Invoke-RestMethod -ParameterFilter { ($Method -eq "Post") -and ($Uri -eq $configurationVersionsUrl) } {
    $null = & $MockForceNonVerboseInvokeRestMethod @PesterBoundParameters
    return @{
        data = @{
            id = $newConfigurationId
            attributes = @{
                status       = "pending"
                "upload-url" = $configurationUploadUrl
            }
        }
    }
}
```

I wasn't expecting it to work - but it did!

Is this expected? Supported? Should I use it (not much of a DRY code improvement in this particular case, but the "default" mock could contain more complex code).

**Comment by fflaten** [comment, marked as answer]
As long as `$MockForceNonVerboseInvokeRestMethod` is created (or dot-sourced) into the same module state where you create the Mock (call the `Mock`-command) it should work IIRC. 

In other words, if you **create** any mocks inside a `InModuleScope ... { }` block, you'd have to modify this a bit by passing the scriptblock-variable into the module, modifying `MockForceNonVerboseInvokeRestMethod` to use a parameter and not the inhertied `$PesterBoundParameters` variable etc.

Personally I'd write a simple MyTestHelpers-module that you'd import before running `Invoke-Pester` or inside a `BeforeDiscovery`-block. Then that "base mock" would be a function with parameters for all required variables (to avoid some module scope issues)
